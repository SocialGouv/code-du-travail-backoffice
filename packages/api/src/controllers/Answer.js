// @ts-check

const AnswerService = require("../services/Answer");

const answerWithError = require("../helpers/answerWithError");
const { COMMON_HEADERS } = require("../constants");

class Answer {
  /**
   * @param {import("http").IncomingMessage} req
   * @param {import("http").ServerResponse} res
   */
  async index(req, res) {
    try {
      const body = await AnswerService.get();

      res.writeHead(200, COMMON_HEADERS);
      res.end(JSON.stringify(body));
    } catch (err) {
      answerWithError("controllers/Answer#index()", err, res);
    }
  }
}

module.exports = new Answer();
