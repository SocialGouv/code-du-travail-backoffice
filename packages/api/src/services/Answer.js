// @ts-check

const knex = require("knex");

const cache = require("../helpers/cache");

const CACHE_TTL = 60 * 60; // => 1h

const { DEV_DB_PORT, NODE_ENV, POSTGRES_DB, POSTGRES_PASSWORD, POSTGRES_USER } = process.env;
let { DB_URI } = process.env;
if (NODE_ENV !== "production") {
  DB_URI = `postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@localhost:${DEV_DB_PORT}/${POSTGRES_DB}`;
}

const knexClient = knex({
  client: "pg",
  connection: DB_URI,
});

/**
 * Get all the publishable answers from the database.
 */
async function get() {
  const cacheKey = "answers";

  // Use cache instead of require if it exists:
  const maybeCachedAnswers = cache.get(cacheKey);
  if (maybeCachedAnswers !== undefined) {
    return maybeCachedAnswers;
  }

  /** @type {Agreement.Agreement[]} */
  const agreements = await knexClient("api.agreements").select();
  /** @type {Answer.Reference[]} */
  const answersReferences = await knexClient("api.answers_references").select();
  /** @type {Question.Question[]} */
  const questions = await knexClient("api.questions").select().limit(10);

  /** @type {Answer.Answer[]} */
  const answers = await knexClient("api.answers")
    .where({
      is_published: true,
      state: "validated",
    })
    .limit(10)
    .select();

  const answersWithAgreement = answers.map(({ agreement_id, ...props }) => ({
    ...props,
    agreement: agreement_id !== null ? agreements.find(({ id }) => id === agreement_id) : null,
  }));
  const answersWithQuestion = answersWithAgreement.map(({ question_id, ...props }) => ({
    ...props,
    question: question_id !== null ? questions.find(({ id }) => id === question_id) : null,
  }));
  const answersWithReferences = answersWithQuestion.map(answer => ({
    ...answer,
    references: answersReferences.filter(({ answer_id }) => answer_id === answer.id),
  }));

  cache.set(cacheKey, answersWithReferences, CACHE_TTL);

  return answersWithReferences;
}

module.exports = {
  get,
};
