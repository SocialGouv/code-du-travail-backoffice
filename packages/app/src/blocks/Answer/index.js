import styled from "@emotion/styled";
import React from "react";
import { Flex } from "rebass";

import { ANSWER_GENERIC_REFERENCE, ANSWER_STATE } from "../../constants";
import Idcc from "../../elements/Idcc";
import Link from "../../elements/Link";

const Content = styled(Flex)`
  background-color: white;
  border: solid 1px var(--color-border);
  border-radius: 0.4rem;
  cursor: pointer;
  flex-grow: 1;
  margin: 0.5rem 0;
  padding: 0.75rem 0.5rem 0.75rem 0.75rem;
  user-select: none;
`;
const ContentQuestion = styled.div`
  color: black;
  font-weight: 600;
`;
const ContentExtract = styled.div`
  color: var(--color-text-gray);
  font-size: 0.875rem;
  margin-top: 0.5rem;
`;
const ContentExtractRed = styled(ContentExtract)`
  color: var(--color-text-red);
  font-weight: 700;
`;

const Menu = styled(Flex)`
  font-weight: 600;
  padding-left: 1.5rem;
  min-width: 13rem;
`;

const AnswerBlock = ({ data, onCancel, onClick, onFallback }) => {
  const isDraft = data.state === ANSWER_STATE.DRAFT;
  const isTodo = data.state === ANSWER_STATE.TO_DO;
  const value = isDraft ? data.prevalue : data.value;

  return (
    <Flex width={1}>
      <Content flexDirection="column" onClick={() => onClick(data.id)}>
        <Flex alignItems="baseline">
          {data.agreement !== null ? (
            <Idcc code={data.agreement_idcc} name={data.agreement_name} />
          ) : (
            <Idcc code="CDT" name="Code du travail" />
          )}
          <ContentQuestion>{`${data.question_index}) ${data.question_value}`}</ContentQuestion>
        </Flex>
        {!isTodo && data.generic_reference === null && (
          <ContentExtract>{value.substr(0, 100)}…</ContentExtract>
        )}
        {!isTodo && data.generic_reference === ANSWER_GENERIC_REFERENCE.LABOR_CODE && (
          <ContentExtractRed>Renvoyé au Code du travail.</ContentExtractRed>
        )}
        {!isTodo && data.generic_reference === ANSWER_GENERIC_REFERENCE.NATIONAL_AGREEMENT && (
          <ContentExtractRed>Renvoyé à la convention collective nationale.</ContentExtractRed>
        )}
      </Content>
      <Menu alignItems="start" flexDirection="column" justifyContent="space-evenly">
        {isTodo && [
          <Link
            disabled={data.generic_reference === ANSWER_GENERIC_REFERENCE.LABOR_CODE}
            isSmall
            key="0"
            onClick={() => onFallback(data.id, ANSWER_GENERIC_REFERENCE.LABOR_CODE)}
          >
            Renvoi au Code du travail
          </Link>,
          <Link
            disabled={data.generic_reference === ANSWER_GENERIC_REFERENCE.NATIONAL_AGREEMENT}
            isSmall
            key="1"
            onClick={() => onFallback(data.id, ANSWER_GENERIC_REFERENCE.NATIONAL_AGREEMENT)}
          >
            Renvoi à la CC nationale
          </Link>,
        ]}
        {isDraft && (
          <Link isSmall onClick={() => onCancel(data.id)}>
            Annuler ma réponse
          </Link>
        )}
      </Menu>
    </Flex>
  );
};

export default AnswerBlock;
