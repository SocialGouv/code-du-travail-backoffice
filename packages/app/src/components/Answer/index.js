import PropTypes from "prop-types";
import React from "react";
import { Flex } from "rebass";

import { ANSWER_STATE, ANSWER_STATE_LABEL } from "../../constants";
import Checkbox from "../../elements/Checkbox";
import Idcc from "../../elements/Idcc";
import excerpt from "../../helpers/excerpt";
import customMoment from "../../libs/customMoment";
import FullAnswerProps from "../../props/FullAnswer";
import {
  Container,
  Content,
  Extract,
  IsPublished,
  Question,
  State,
  Top,
  UpdatedAt,
} from "./index.style";

const Answer = ({ data, isChecked, onCheck, onClick }) => {
  const {
    agreement_idcc,
    agreement_name,
    generic_reference,
    id,
    is_published,
    question_index,
    question_value,
    state,
    updated_at,
  } = data;

  const value = [ANSWER_STATE.DRAFT, ANSWER_STATE.PENDING_REVIEW].includes(state)
    ? data.prevalue
    : data.value;

  if (state === ANSWER_STATE.TO_DO) {
    return (
      <Container flexDirection="column">
        <Top justifyContent="space-between">
          <State>{ANSWER_STATE_LABEL[state]}</State>
        </Top>
        <Flex>
          <Checkbox
            isChecked={isChecked}
            onClick={() => onCheck(id)}
            withMarginLeft
            withMarginRight
          />
          <Content flexDirection="column" onClick={() => onClick(id)}>
            <Flex alignItems="baseline">
              <Idcc code={agreement_idcc} name={agreement_name} />
              <Question>{`${question_index}) ${question_value}`}</Question>
            </Flex>
          </Content>
        </Flex>
      </Container>
    );
  }

  return (
    <Container flexDirection="column">
      <Top justifyContent="space-between">
        <State>{ANSWER_STATE_LABEL[state]}</State>
        <Flex alignItems="center">
          <UpdatedAt>{`Modifié ${customMoment(updated_at)
            .tz("Europe/Paris")
            .calendar()}`}</UpdatedAt>
          <IsPublished isDisabled={!is_published}>Publiée</IsPublished>
        </Flex>
      </Top>
      <Flex>
        <Checkbox
          isChecked={isChecked}
          onClick={() => onCheck(id)}
          withMarginLeft
          withMarginRight
        />
        <Content flexDirection="column" onClick={() => onClick(id)}>
          <Flex alignItems="baseline">
            <Idcc code={agreement_idcc} name={agreement_name} />
            <Question>{`${question_index}) ${question_value}`}</Question>
          </Flex>
          {generic_reference === null && <Extract>{excerpt(value)}</Extract>}
          {generic_reference === "labor_code" && (
            <Extract isGeneric>Renvoyé au Code du travail.</Extract>
          )}
          {generic_reference === "national_agreement" && (
            <Extract isGeneric>Renvoyé à la convention collective nationale.</Extract>
          )}
        </Content>
      </Flex>
    </Container>
  );
};

Answer.propTypes = {
  data: PropTypes.shape(FullAnswerProps),
  isChecked: PropTypes.bool.isRequired,
  onCheck: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default Answer;
