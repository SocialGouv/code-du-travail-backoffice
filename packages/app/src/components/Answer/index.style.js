import styled from "@emotion/styled";
import { Flex } from "rebass";

export const Container = styled(Flex)`
  margin-top: 1rem;
  user-select: none;
`;

export const Top = styled(Flex)`
  font-size: 0.75rem;
  font-weight: 700;
  margin-bottom: 0.25rem;
  padding-left: 3.25rem;
  text-transform: uppercase;
`;

export const State = styled.span`
  color: var(--color-shadow);
`;

export const UpdatedAt = styled.span`
  color: var(--color-blue-sapphire);
`;

export const Content = styled(Flex)`
  background-color: white;
  border: solid 1px var(--color-border);
  border-radius: 0.4rem;
  cursor: pointer;
  flex-grow: 1;
  padding: 0.75rem 0.5rem 0.75rem 0.75rem;
`;

export const Question = styled.div`
  color: black;
  font-weight: 600;
`;

export const Extract = styled.div`
  color: ${p => (p.isGeneric ? "var(--color-text-red)" : "var(--color-text-gray)")};
  font-size: 0.875rem;
  font-weight: ${p => (p.isGeneric ? "700" : "normal")};
  margin-top: 0.5rem;
`;
