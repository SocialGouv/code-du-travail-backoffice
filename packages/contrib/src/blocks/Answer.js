import React from "react";
import { Flex } from "rebass";
import styled from "styled-components";

import Idcc from "../elements/Idcc";

const Container = styled.div`
  background-color: white;
  border-radius: 0.4rem;
  cursor: pointer;
  margin: 1rem 1rem 0;
  padding: 0.5rem 0.5rem 0.5rem 0.75rem;
`;

const Label = styled.div`
  color: #c2c2c2;
  font-size: 0.8rem;
  font-weight: 600;
`;

export default ({ data, label, onClick }) => (
  <Container onClick={onClick}>
    <Flex justifyContent="space-between">
      <Label>{label}</Label>
      <Idcc code={data.idcc} name={data.agreement} />
    </Flex>
    {data.question}
  </Container>
);
