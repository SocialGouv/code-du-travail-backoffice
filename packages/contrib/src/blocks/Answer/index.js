import React from "react";
import { Flex } from "rebass";
import styled from "styled-components";

import Idcc from "../../elements/Idcc";
import Link from "../../elements/Link";

import { ANSWER_STATE, ANSWER_GENERIC_REFERENCE } from "../../constants";

const Content = styled(Flex)`
  background-color: white;
  border: solid 1px var(--color-border);
  border-radius: 0.4rem;
  cursor: ${({ isEditable }) => (isEditable ? "pointer" : "auto")};
  flex-grow: 1;
  margin: 0.5rem 0;
  padding: 0.75rem 0.5rem 0.75rem 0.75rem;
  user-select: none;
`;
const ContentQuestion = styled.div`
  color: black;
  font-weight: 600;
`;
const ContentExtract = styled.div`
  color: var(--color-text-gray);
  font-size: 0.875rem;
  margin-top: 0.5rem;
`;
const ContentExtractRed = styled(ContentExtract)`
  color: var(--color-text-red);
  font-weight: 700;
`;

const Menu = styled(Flex)`
  font-weight: 600;
  padding-left: 1.5rem;
  min-width: 13rem;
`;

export default ({ data, onCancel, onClick, onFallback }) => {
  const isDraft = data.state === ANSWER_STATE.DRAFT;
  const isTodo = data.state === ANSWER_STATE.TO_DO;
  const value = isDraft ? data.prevalue : data.value;

  const isEditable = isTodo || isDraft;
  const _onClick = () => (isEditable ? onClick(data.id) : void 0);

  return (
    <Flex width={1}>
      <Content
        flexDirection="column"
        isEditable={isEditable}
        onClick={_onClick}
      >
        <Flex alignItems="baseline">
          <Idcc code={data.idcc} name={data.agreement} />
          <ContentQuestion>{`${data.index}) ${data.question}`}</ContentQuestion>
        </Flex>
        {!isTodo && data.generic_reference === null && (
          <ContentExtract>{value.substr(0, 100)}…</ContentExtract>
        )}
        {!isTodo &&
          data.generic_reference === ANSWER_GENERIC_REFERENCE.LABOR_CODE && (
            <ContentExtractRed>Renvoyé au Code du travail.</ContentExtractRed>
          )}
        {!isTodo &&
          data.generic_reference ===
            ANSWER_GENERIC_REFERENCE.NATIONAL_AGREEMENT && (
            <ContentExtractRed>
              Renvoyé à la convention collective nationale.
            </ContentExtractRed>
          )}
      </Content>
      <Menu
        alignItems="start"
        justifyContent="space-evenly"
        flexDirection="column"
      >
        {isTodo && [
          <Link
            disabled={
              data.generic_reference === ANSWER_GENERIC_REFERENCE.LABOR_CODE
            }
            key="0"
            isSmall
            onClick={() =>
              onFallback(data.id, ANSWER_GENERIC_REFERENCE.LABOR_CODE)
            }
          >
            Renvoi au Code du travail
          </Link>,
          <Link
            disabled={
              data.generic_reference ===
              ANSWER_GENERIC_REFERENCE.NATIONAL_AGREEMENT
            }
            key="1"
            isSmall
            onClick={() =>
              onFallback(data.id, ANSWER_GENERIC_REFERENCE.NATIONAL_AGREEMENT)
            }
          >
            Renvoi à la CC nationale
          </Link>
        ]}
        {isDraft && (
          <Link isSmall onClick={() => onCancel(data.id)}>
            Annuler ma réponse
          </Link>
        )}
      </Menu>
    </Flex>
  );
};
