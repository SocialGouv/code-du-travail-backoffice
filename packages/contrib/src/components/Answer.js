import React from "react";
import ReactTooltip from "react-tooltip";
import { Flex } from "rebass";
import styled from "styled-components";

const Container = styled.div`
  background-color: white;
  border-radius: 0.4rem;
  cursor: pointer;
  margin: 1rem 1rem 0;
  padding: 0.5rem 0.5rem 0.5rem 0.75rem;
`;

const Label = styled.div`
  color: #c2c2c2;
  font-size: 0.8rem;
  font-weight: 600;
`;
const Reference = styled.div`
  background-color: #606060;
  color: white;
  font-size: 0.8rem;
  font-weight: 600;
  opacity: 0.75;
  padding: 0.25rem 0.5rem;

  :hover {
    opacity: 1;
  }
`;
const Tooltip = styled(ReactTooltip)`
  padding: 0.25rem 0.5rem 0.45rem;
  width: 360px;
`;

export default ({ data, label, onClick }) => (
  <Container key={data.id} onClick={onClick}>
    <Flex justifyContent="space-between">
      <Label>{label}</Label>
      <Reference data-tip={data.labor_agreement.name}>
        {data.labor_agreement.idcc}
      </Reference>
      <Tooltip />
    </Flex>
    {data.question.value}
  </Container>
);
