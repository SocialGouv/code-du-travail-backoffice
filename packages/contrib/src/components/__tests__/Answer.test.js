import React from "react";
import { fireEvent, render } from "react-testing-library";
import Answer from "../Answer";

describe("<Answer />", () => {
  const onClick = jest.fn();
  const props = {
    data: {
      id: "12345678-9abc-4def-0123-456789abcdef",
      labor_agreement: {
        idcc: "1234",
        name: "A Labor Agreement Name"
      },
      question: {
        value: "Who knows?"
      }
    },
    label: "A label",
    onClick
  };
  const { container, getByText } = render(<Answer id="1" {...props} />);

  it("should match snapshot", () => {
    expect(container).toMatchSnapshot();
  });

  describe("when clicked twice", () => {
    const $ = getByText(/Who knows/);
    fireEvent.click($);
    fireEvent.click($);

    it("should have called onClick() twice", () => {
      expect(onClick.mock.calls.length).toBe(2);
    });
  });
});
