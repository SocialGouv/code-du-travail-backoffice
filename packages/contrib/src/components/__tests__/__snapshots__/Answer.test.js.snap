// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Answer /> should match snapshot 1`] = `
<div>
  <div
    class="Answer__Container-sc-1w2h0sp-0 kPKGxi"
  >
    <div
      class="sc-bdVaJa sc-bwzfXH erVoqY"
    >
      <div
        class="Answer__Label-sc-1w2h0sp-1 ikeTNC"
      >
        A label
      </div>
      <div
        class="Answer__Reference-sc-1w2h0sp-2 iCNfob"
        currentitem="false"
        data-tip="A Labor Agreement Name"
      >
        1234
      </div>
      <div
        class="__react_component_tooltip place-top type-dark "
        data-id="tooltip"
      />
    </div>
    Who knows?
  </div>
</div>
`;
